libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/kpm/external/ExternalVariantApi.kt

```diff
@@ -33,3 +30,7 @@ fun KotlinGradleModule.createExternalJvmVariant(
 @ExternalVariantApi
 val KotlinGradleVariantInternal.compilationData
     get() = this.compilationData
+
+@ExternalVariantApi
+val KotlinGradleFragment.external: KotlinMutableExternalModelContainer
```

Comment:
IMHO an API where there is just a registry of `AdditionalDataProvider<T, Key<T>>` may express the intention in a nicer way. Then, the registry is realized during model building and adding new providers will fail (as a safety check).